### PR TITLE
Backporting shape hash optimization

### PR DESCRIPTION
**Pull request by acunniffe**

### CODE REVIEW

File: core/domain-snapshots/events-interactions-diff-interpretation-ui-render/outputs/a required array field of strings provided with an object.managed.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
                   },
                   {
                     "JsonArrayItem" : {
-                      "index" : 0
+                      "index" : 1
```
Comment (devdoshi) [inline core/domain-snapshots/events-interactions-diff-interpretation-ui-render/outputs/a required array field of strings provided with an object.managed.json:32]: did we establish why this one is now 1?

File: core/domain-snapshots/events-interactions-diff-interpretation-ui-render/outputs/when a nullable is provided with a concrete type.managed.json [hunk 2]
```diff
@@ -24,7 +24,7 @@
                 "path" : [
                   {
                     "JsonArrayItem" : {
-                      "index" : 0
+                      "index" : 1
```
Comment (devdoshi) [inline core/domain-snapshots/events-interactions-diff-interpretation-ui-render/outputs/when a nullable is provided with a concrete type.managed.json:27]: same question here

File: core/optic/shared/src/main/scala/com/useoptic/types/capture/JsonLike.scala [hunk 3]
```diff
@@ -22,45 +23,49 @@ abstract class JsonLike {
...
-
-    override def isObject: Boolean = false
-
-    override def fields: Map[String, JsonLike] = Map.empty
-
-    override def items: Vector[JsonLike] = a
-
-    override def asJson: Json = Json.fromValues(a.map(_.asJson))
-  }
+//  def map(value: Map[String, JsonLike]): JsonLike = new JsonLike {
```
Comment (devdoshi) [inline core/optic/shared/src/main/scala/com/useoptic/types/capture/JsonLike.scala:36]: Can we delete this?

File: core/optic/shared/src/test/scala/com/useoptic/end_to_end/DiffUseCasesSpec.scala [hunk 4]
```diff
@@ -59,6 +59,8 @@ class DiffUseCasesSpec extends EndEndDiffTask with JsonFileFixture {
           Json.fromString("San Fransisco"),
           Json.fromInt(17584),
           Json.fromString("Boston"),
+          Json.fromString("California"),
```
Comment (devdoshi) [inline core/optic/shared/src/test/scala/com/useoptic/end_to_end/DiffUseCasesSpec.scala:62]: I think the following test covers the scenario, right? so we can revert this

**Comment by acunniffe** [comment]
@devdoshi well turns out the thing we did before didn't work right. using sets now and everything seems friendly again

**Pull request merged**